core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -36,373 +73,288 @@ class Queue[A] private (
-
-    val acquire: (Promise[Nothing, A], State[A]) => (IO[Nothing, Boolean], State[A]) = {
-      case (p, Deficit(takers, hook)) => (IO.now(false), Deficit(takers.enqueue(p), hook))
-      case (p, Surplus(values, putters, hook)) =>
-        strategy match {
-          case Sliding | Dropping if capacity.exists(_ < 1) =>
-            (IO.never, Surplus(IQueue.empty, putters, hook))
-          case _ =>
-            values.dequeueOption match {
-              case None if putters.isEmpty =>
-                (IO.now(false), Deficit(IQueue.empty.enqueue(p), hook))
-              case None =>
-                val (newSurplus, promise) =
-                  moveNPutters(Surplus(values, putters, hook), 1)
-                newSurplus.queue.dequeueOption match {
-                  case None => (promise *> IO.now(false), newSurplus)
-                  case Some((a, values)) =>
-                    (
-                      promise *> p.complete(a),
-                      Surplus(values, newSurplus.putters, hook)
-                    )
-                }
-              case Some((a, values)) =>
-                val (newSurplus, promise) =
-                  moveNPutters(Surplus(values, putters, hook), 1)
-                (promise *> p.complete(a), newSurplus)
-
-            }
-        }
-      case (p, state @ Shutdown) => (p.interrupt, state)
-    }
-
-    val release: (Boolean, Promise[Nothing, A]) => IO[Nothing, Unit] = {
-      case (_, p) => p.poll.void <> removeTaker(p)
-    }
-
-    Promise.bracket[Nothing, State[A], A, Boolean](ref)(acquire)(release)
-
-  }
+  final val take: IO[Nothing, A] =
+    for {
+      _    <- checkShutdownState
+      item <- IO.sync(queue.poll(null.asInstanceOf[A]))
+      res <- if (item != null) strategy.onQueueEmptySpace(queue).map(_ => item)
+            else
+              for {
+                p <- Promise.make[Nothing, A]
+                // add the promise to takers, then try take again in case a value was added since
+                _ <- IO.sync(takers.offer(p)) *> completeTakers
+                // wait for the promise to be completed, and clean up resources in case of interruption
+                res <- p.get.ensuring(p.poll.void <> removeTaker(p))
+              } yield res
+    } yield res
 
   /**
    * Removes all the values in the queue and returns the list of the values. If the queue
    * is empty returns empty list.
    */
   final val takeAll: IO[Nothing, List[A]] =
-    IO.flatten(ref.modify[IO[Nothing, List[A]]] {
-      case Surplus(values, putters, hook) =>
-        val (newState, promises) = moveNPutters(Surplus(IQueue.empty, putters, hook), values.size)
-        (promises *> IO.point(values.toList), newState)
-      case state @ Deficit(_, _) => (IO.point(List.empty[A]), state)
-      case state @ Shutdown      => (IO.interrupt, state)
-    })
+    for {
+      _   <- checkShutdownState
+      res <- IO.sync(unsafePollAll(queue))
+      _   <- strategy.onQueueEmptySpace(queue)
```

Comment:
Something to think about is what happens when these actions are interrupted mid-way through? i.e. after `unsafePollAll` but before `strategy.onQueueEmptySpace`?

Reply:
Yeah there might be some unexpected behavior in case of interruption (e.g. a taker interrupted just before it was going to complete a putter, or vice-versa).

Should we use `uninterruptibly` to prevent the instantaneous actions to be interrupted? I imagine only the `Promise.get` would need to be interrupted.

Reply:
I can capture the current `executionContext` thanks to `IO.descriptor` and re-use it to complete the promises with `unsafeDone`. Performance is great even though the code is a bit more complex. Does this make any sense?

Reply:
@ghostdogpr 

Yes, it does make sense. Always push as much code as you can out of `IO`, and try to only wrap once in `sync`. That will ensure optimum performance. Also: code inside a `sync` cannot be interrupted, so you don't have to use the (much slower) `uninterruptible` combinator.

Once we get this merged, I can help walk through the code and show you some allocations that can be eliminated to improve performance. It will be very low-level optimization, though, and we should have good benchmarks merged in before we begin that process.